File: src/main/java/ui/pageobjects/pages/SearchResultsPage.java
package ui.pageobjects.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.LoadState;
import ui.pageobjects.fragments.searchresults.VideoResultItem;

import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class SearchResultsPage extends BasePage {
    public SearchResultsPage(Page page) {
        super(page);
    }

    public Locator resultsCounter() {
        return this.page.locator(".results_count");
    }

    public List<VideoResultItem> videoResultItems() {
        Locator searchResultItems = searchResultItems();
        return searchResultItems.all().stream().map(wrapper -> new VideoResultItem(this.page, wrapper)).collect(Collectors.toList());
    }

    public Locator searchResultItems() {
        return this.page.locator(".iris_p_infinite__item");
    }

    public Locator paginationDots() {
        return this.getPaginationWrapper().locator(".iris_pagination__dots");
    }

    public Locator paginationNextButton() {
        return this.getPaginationWrapper().getByText("Next");
    }

    public Locator paginationPrevButton() {
        return this.getPaginationWrapper().getByText("Prev");
    }

    public Locator paginationButton(int pageNumber) {
        return this.paginationNumberButtons().getByText(String.valueOf(pageNumber));
    }

    public Locator paginationNumberButtons() {
        return this.getPaginationWrapper().locator(".iris_pagination__list-item")
                .filter(new Locator.FilterOptions().setHasNotText(Pattern.compile("Prev|Next")));
    }

    public SearchResultsPage clickFilter(String label) {
        Locator filterButton = this.page.locator(".iris_rail-filter__list-item", new Page.LocatorOptions().setHasText(label));
        waitForSearchResults(() -> filterButton.click());
        return this;
    }

    public SearchResultsPage clickPagination(int pageNumber) {
        waitForSearchResults(() -> this.paginationButton(pageNumber).click());
        return this;
    }

    public SearchResultsPage clickLastPagination() {
        waitForSearchResults(() -> this.paginationNumberButtons().last().click());
        return this;
    }

    public SearchResultsPage clickNext() {
        waitForSearchResults(() -> this.paginationNextButton().click());
        return this;
    }

    public SearchResultsPage clickPrevious() {
        waitForSearchResults(() -> this.paginationPrevButton().click());
        return this;
    }

    void waitForSearchResults(Runnable callback) {
        this.page.waitForResponse(response -> response.url().contains("/search"), callback);
        this.page.waitForLoadState(LoadState.DOMCONTENTLOADED);
    }

    private Locator getPaginationWrapper() {
        return this.page.locator("#pagination");
    }
}

File: src/main/kotlin/ui/pageobjects/pages/SearchResultsPage.kt
package ui.pageobjects.pages

import com.microsoft.playwright.Locator
import com.microsoft.playwright.Locator.FilterOptions
import com.microsoft.playwright.Page
import com.microsoft.playwright.Response
import com.microsoft.playwright.options.LoadState
import ui.pageobjects.fragments.searchresults.VideoResultItem
import java.util.regex.Pattern
import java.util.stream.Collectors

class SearchResultsPage(page: Page?) : BasePage(page!!) {
    fun resultsCounter(): Locator {
        return page.locator(".results_count")
    }

    fun videoResultItems(): List<VideoResultItem> {
        val searchResultItems = searchResultItems()
        return searchResultItems.all().stream().map { wrapper: Locator? -> VideoResultItem(this.page, wrapper!!) }.collect(Collectors.toList())
    }

    fun searchResultItems(): Locator {
        return page.locator(".iris_p_infinite__item")
    }

    fun paginationDots(): Locator {
        return paginationWrapper.locator(".iris_pagination__dots")
    }

    fun paginationNextButton(): Locator {
        return paginationWrapper.getByText("Next")
    }

    fun paginationPrevButton(): Locator {
        return paginationWrapper.getByText("Prev")
    }

    fun paginationButton(pageNumber: Int): Locator {
        return paginationNumberButtons().getByText(pageNumber.toString())
    }

    fun paginationNumberButtons(): Locator {
        return paginationWrapper.locator(".iris_pagination__list-item")
                .filter(FilterOptions().setHasNotText(Pattern.compile("Prev|Next")))
    }

    fun clickFilter(label: String?): SearchResultsPage {
        val filterButton = page.locator(".iris_rail-filter__list-item", Page.LocatorOptions().setHasText(label))
        waitForSearchResults { filterButton.click() }
        return this
    }

    fun clickPagination(pageNumber: Int): SearchResultsPage {
        waitForSearchResults { paginationButton(pageNumber).click() }
        return this
    }

    fun clickLastPagination(): SearchResultsPage {
        waitForSearchResults { paginationNumberButtons().last().click() }
        return this
    }

    fun clickNext(): SearchResultsPage {
        waitForSearchResults { paginationNextButton().click() }
        return this
    }

    fun clickPrevious(): SearchResultsPage {
        waitForSearchResults { paginationPrevButton().click() }
        return this
    }

    fun waitForSearchResults(callback: Runnable?) {
        page.waitForResponse({ response: Response -> response.url().contains("/search") }, callback)
        page.waitForLoadState(LoadState.DOMCONTENTLOADED)
    }

    private val paginationWrapper: Locator
        get() = page.locator("#pagination")
}
